accepted/2021/workloads/workload-target-imports.md

```diff
@@ -0,0 +1,118 @@
+
+## Proposed Solutions
+
+The following are the current proposed solutions to the problem outlined above, to be reviewed by the community.
+
+### Extension Point via Property
+
+Support an `AfterTargetFrameworkInferenceTargets` property. This property could be used by creating a `Directory.AfterTargetFrameworkInference.targets` file and putting the following in the `Directory.Build.props` file located in the same folder:
+
+```xml
+<AfterTargetFrameworkInferenceTargets>$(MSBuildThisFileDirectory)Directory.AfterTargetFrameworkInference.targets</AfterTargetFrameworkInferenceTargets>
+```
+
+#### Pros:
+
+- Simple to implement
+- Low compat and performance impacts
+- Matches existing `BeforeTargetFrameworkInferenceTargets` property
+
+#### Cons:
+
+- Doesn’t match existing `Directory.Build.props` and `Directory.Build.targets` pattern, or the general principle of the SDK which is to have sensible convention-based defaults that can be overridden
+
+### New Automatically Imported .targets File
+
+Automatically find and import a `Directory.AfterTargetFrameworkInference.targets` file.
+
+#### Pros:
+
+- Matches existing convention for `Directory.Build.props` and `Directory.Build.targets`
+- Probably closest to what developers would expect
+
+#### Cons:
+
+- Possible performance impact
+- Adds another way for a build to “leak out” of a repo root (which could be a security issue)
+- More extension points may be needed in the future, and we probably don’t want to add a new auto-imported file for each one
+
+A possible mitigation for some of the cons could be to say that we don’t look for the new file to import independently. Rather, we could say that you need to have a `Directory.Build.props` file and it needs to be in the same folder as `Directory.Build.props`. So a `Directory.AfterTargetFrameworkInference.targets` file outside the repo root wouldn’t automatically be imported unless a `Directory.Build.props` file was going to be imported from outside the repo anyway.
+
+### Change import location of `Directory.Build.targets`
+
+Change`Directory.Build.targets` to be imported after TargetFramework parsing but before workloads and most of the .NET SDK and common targets. This would be a big breaking change but might not affect most people who use `Directory.Build.targets`. It would break people who use `Directory.Build.targets` to override properties or targets in the MSBuild common targets.
+
+#### Pros:
+
+- Logic to enable workloads in `Directory.Build.targets` would “just work”, which is likely what developers expect
```

Comment:
I strongly prefer this option. It took a while to evangelize the use of `Directory.Build.props` and `Directory.Build.targets`. I believe we should strive to optimize for letting developers do 99% of their tweaks by just these files. Having to put more stuff in new files whose import order is slightly different is a complexity nightmare. Yes, this is a breaking change but I think that's a perfect example where the net harm of the breaking change is *much* lower than the additional cognitive load over more files over time.

Reply:
It's hard to grok how significant this is or is not though. Can we get some examples of which properties we can no longer override in `Directory.Build.targets` if we take this approach. At the moment it's pretty nebulous hence it's hard to decide if this is or is not the right approach.